### PR TITLE
Fix #568, fix #572, fix #565

### LINKED ISSUE #565: MockPin and MockPWMPin documentation missing
Any reason `MockPin` and `MockPWMPin` are left out of the [pins page](https://gpiozero.readthedocs.io/en/stable/api_pins.html) in the docs?

It's a shame these are not documented - we should ensure they are included. We should show people the API docs for these classes, and give the environment variables requires to use them by default - and state that they are cross-platform and ideal for testing code on a PC.

### LINKED ISSUE #568: RPiGPIOPin / RPiGPIOFactory confusion
Previously, you could pass an `RPiGPIOPin` object as the pin reference to initialise a GPIO object like so:

```python
from gpiozero import LED
from gpiozero.pins.rpigpio import RPiGPIOPin

led = LED(RPiGPIOPin(12))
```

(example straight from the [1.3.2 docs](http://gpiozero.readthedocs.io/en/v1.3.2/api_pins.html#rpigpiopin))

However, that doesn't work on the current master. An [updated version of the docs](http://gpiozero.readthedocs.io/en/latest/api_pins.html#rpigpiopin) suggests the following invalid example:

```python
from gpiozero.pins.rpigpio import RPiGPIOFactory
from gpiozero import LED

led = LED(RPiGPIOPin(12))
```

Note the discrepancy: `RPiGPIOFactory` is imported but not used, and `RPiGPIOPin` is referenced but not imported.

I can produce an equivalent example using the following:

```python
from gpiozero.pins.rpigpio import RPiGPIOFactory
from gpiozero import LED

led = LED(RPiGPIOPin(RPiGPIOFactory(), 12))
```

Which is kinda yucky.

@waveform80 what's the recommended way of doing the original example above now?

### LINKED ISSUE #572: Make "failed to load factory" warnings not look like fatal errors
```
pi@raspberrypi:~ $ git clone https://github.com/rpi-distro/python-gpiozero
Cloning into 'python-gpiozero'...
remote: Counting objects: 3669, done.
remote: Compressing objects: 100% (5/5), done.
remote: Total 3669 (delta 0), reused 2 (delta 0), pack-reused 3664
Receiving objects: 100% (3669/3669), 6.82 MiB | 349.00 KiB/s, done.
Resolving deltas: 100% (2450/2450), done.
Checking connectivity... done.

pi@raspberrypi:~ $ virtualenv -p python3 gp0-env
Already using interpreter /usr/bin/python3
Using base prefix '/usr'
New python executable in gp0-env/bin/python3
Also creating executable in gp0-env/bin/python
Installing setuptools, pip...done.

pi@raspberrypi:~ $ source gp0-env/bin/activate

(gp0-env)pi@raspberrypi:~ $ cd python-gpiozero/

(gp0-env)pi@raspberrypi:~/python-gpiozero $ python setup.py develop
running develop
running egg_info
creating gpiozero.egg-info
writing gpiozero.egg-info/PKG-INFO
writing requirements to gpiozero.egg-info/requires.txt
writing dependency_links to gpiozero.egg-info/dependency_links.txt
writing entry points to gpiozero.egg-info/entry_points.txt
writing top-level names to gpiozero.egg-info/top_level.txt
writing manifest file 'gpiozero.egg-info/SOURCES.txt'
reading manifest file 'gpiozero.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
writing manifest file 'gpiozero.egg-info/SOURCES.txt'
running build_ext
Creating /home/pi/gp0-env/lib/python3.4/site-packages/gpiozero.egg-link (link to .)
Adding gpiozero 1.3.2 to easy-install.pth file
Installing pinout script to /home/pi/gp0-env/bin

Installed /home/pi/python-gpiozero
Processing dependencies for gpiozero==1.3.2
Finished processing dependencies for gpiozero==1.3.2

(gp0-env)pi@raspberrypi:~/python-gpiozero $ pip install pigpio
Downloading/unpacking pigpio
  Downloading pigpio-1.35.tar.gz
  Running setup.py (path:/tmp/pip-build-n1_fsqv1/pigpio/setup.py) egg_info for package pigpio
    
Installing collected packages: pigpio
  Running setup.py install for pigpio
    
Successfully installed pigpio
Cleaning up...

(gp0-env)pi@raspberrypi:~/python-gpiozero $ python3
Python 3.4.2 (default, Oct 19 2014, 13:31:11) 
[GCC 4.9.1] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from gpiozero import LED
/home/pi/python-gpiozero/gpiozero/devices.py:505: PinFactoryFallback: Failed to load factory rpigpio: No module named 'RPi'
  'Failed to load factory %s: %s' % (name, str(e))))
/home/pi/python-gpiozero/gpiozero/devices.py:505: PinFactoryFallback: Failed to load factory rpio: No module named 'RPIO'
  'Failed to load factory %s: %s' % (name, str(e))))
>>> 
```

### PR DESCRIPTION
**Pull request by waveform80**
Update examples in pin factory docs so they work correctly.

### CODE REVIEW

File: docs/examples/led_button_remote_2.py [hunk 1]
```diff
@@ -1,11 +1,14 @@
 from gpiozero import LED
-from gpiozero.pins.pigpio import PiGPIOPin
+from gpiozero.pins.pigpio import PiGPIOFactory
 from gpiozero.tools import all_values
 from signal import pause
 
+factory3 = PiGPIOFactory(host='192.168.1.3')
+factory4 = PiGPIOFactory(host='192.168.1.3')
```
Comment (lurch) [inline docs/examples/led_button_remote_2.py]: Probably want different IPs here ;-)

File: docs/examples/led_remote_2.py [hunk 2]
```diff
@@ -1,9 +1,11 @@
...
-from gpiozero.pins.pigpio import PiGPIOPin
+from gpiozero.pins.pigpio import PiGPIOFactory
 from time import sleep
 
-led_1 = LED(PiGPIOPin(17, host='192.168.1.3'))
-led_2 = LED(PiGPIOPin(17, host='192.168.1.4'))
+factory3 = PiGPIOFactory(host='192.168.1.3')
+factory4 = PiGPIOFactory(host='192.168.1.4')
+led_1 = LED(factory3.pin(17)
+led_2 = LED(factory4.pin(17)
```
Comment (lurch) [inline docs/examples/led_remote_2.py]: Missing closing-brackets

File: docs/examples/led_button_remote_2.py (hunk 1 above)
Comment (waveform80) [inline docs/examples/led_button_remote_2.py]: Ooops :)

File: docs/examples/led_remote_2.py (hunk 2 above)
Comment (waveform80) [inline docs/examples/led_remote_2.py]: Bah - this is what comes of coding while simultaneously trying to discuss technical stuff on the phone!

**Comment by bennuttall** [comment]
I've pushed @waveform80's branch to this repo (branch [pin-factory-docs](https://github.com/RPi-Distro/python-gpiozero/tree/pin-factory-docs)) so that we can browse the docs here: http://gpiozero.readthedocs.io/en/pin-factory-docs/api_pins.html

Note that any changes to Dave's branch (the PR) will need to be merged to this repo's branch.

**Comment by bennuttall** [comment]
Just testing. This is working well for me:

```python
import gpiozero
from gpiozero import TrafficHat
from gpiozero.pins.pigpio import PiGPIOFactory

one = PiGPIOFactory('172.18.62.68')
two = PiGPIOFactory('172.18.64.171')  # yes, they're on different subnets

local_th = TrafficHat()
gpiozero.Device.pin_factory = one
th_one = TrafficHat()
gpiozero.Device.pin_factory = two
th_two = TrafficHat()

local_th.lights.blink()
th_one.lights.blink()
th_two.lights.blink()
```

**Comment by lurch** [comment]
Looking at Ben's example above, where the identical-looking `TrafficHat()` calls do three separate things, I guess it relies on "implicit global state", which is something @waveform80 has ranted about previously...? ;-)

iMHO the proposed (in #279 ) syntax below "feels" much cleaner :)
```python
import gpiozero
from gpiozero import TrafficHat
from gpiozero.pins.pigpio import PiGPIOFactory

one = PiGPIOFactory('172.18.62.68')
two = PiGPIOFactory('172.18.64.171')  # yes, they're on different subnets

local_th = TrafficHat()
th_one = TrafficHat(pin_factory=one)
th_two = TrafficHat(pin_factory=two)

local_th.lights.blink()
th_one.lights.blink()
th_two.lights.blink()
```


Another crazy @lurch comment, but once this "new syntax" is adopted, what if `gpiozero.Device.pin_factory` wasn't publicly settable, but there was instead a `gpiozero.Device.set_default_pin_factory(factory)` function, and then that function could issue a warning if the same program tried to change the default_pin_factory more than once? (because it might lead to a novice-user's code doing "unexpected" things?)

**Comment by lurch** [comment]
> we can browse the docs here: http://gpiozero.readthedocs.io/en/pin-factory-docs/api_pins.html

Looking at that page, does it make sense to include `gpiozero.pins.rpigpio.RPiGPIOPi` and `gpiozero.pins.rpio.RPIOPin` (etc.) given that we don't expect them to be directly instantiated by the user?

Even though it's used in lots of examples, http://gpiozero.readthedocs.io/en/pin-factory-docs/api_pins.html#gpiozero.Factory.pin seems to be the only documentation of the `pin` method?

That page has a singular mention of `i2C` - perhaps it should be removed, or it should be mentioned that it isn't supported yet?

**Comment by bennuttall** [comment]
> Looking at that page, does it make sense to include `gpiozero.pins.rpigpio.RPiGPIOPin` and `gpiozero.pins.rpio.RPIOPin` (etc.) given that we don't expect them to be directly instantiated by the user?

Didn't @waveform80 say he wanted to hide those now?

> That page has a singular mention of i2C - perhaps it should be removed, or it should be mentioned that it isn't supported yet?

Probably :+1:

**Comment by waveform80** [comment]
> That page has a singular mention of i2C - perhaps it should be removed, or it should be mentioned that it isn't supported yet?

Oh good point, I'd missed that.

> Didn't @waveform80 say he wanted to hide those now?

Erm - I don't think I mentioned hiding the pin classes. Actually I think they *probably* should still be documented given that they are created and used at that level, and those docs are meant to cover what goes on at that level ...

**Comment by bennuttall** [comment]
Ok sure

**Comment by bennuttall** [comment]
Great! May as well merge and fix anything else that comes up on master. I believe we're very close to release now...

**Comment by waveform80** [comment]
> Great! May as well merge and fix anything else that comes up on master. I believe we're very close to release now...

Yup, that's my hope :)

**Comment by bennuttall** [comment]
I've got a few docs tweaks I'm making. You ok to merge this?

**Comment by bennuttall** [comment]
Oh, you just did. Thanks!

**Pull request merged**